server/client.go

```diff
@@ -974,61 +1253,95 @@ func (c *client) deliverMsg(sub *subscription, mh, msg []byte) bool {
 	client.outMsgs++
 	client.outBytes += msgSize
 
-	atomic.AddInt64(&c.srv.outMsgs, 1)
-	atomic.AddInt64(&c.srv.outBytes, msgSize)
-
-	// Check to see if our writes will cause a flush
-	// in the underlying bufio. If so limit time we
-	// will wait for flush to complete.
+	atomic.AddInt64(&srv.outMsgs, 1)
+	atomic.AddInt64(&srv.outBytes, msgSize)
 
-	deadlineSet := false
-	if client.bw.Available() < (len(mh) + len(msg)) {
-		client.wfc++
-		client.nc.SetWriteDeadline(time.Now().Add(client.srv.getOpts().WriteDeadline))
-		deadlineSet = true
-	}
+	// Queue to outbound buffer
+	client.queueOutbound(mh)
+	client.queueOutbound(msg)
 
-	// Deliver to the client.
-	_, err := client.bw.Write(mh)
-	if err != nil {
-		goto writeErr
-	}
+	client.out.pm++
 
-	_, err = client.bw.Write(msg)
-	if err != nil {
-		goto writeErr
+	// Check outbound threshold and queue IO flush if needed.
+	if client.out.pm > 1 && client.out.pb > maxBufSize*2 {
+		client.flushSignal()
 	}
 
 	if c.trace {
 		client.traceOutOp(string(mh[:len(mh)-LEN_CR_LF]), nil)
 	}
 
-	// TODO(dlc) - Do we need this or can we just call always?
-	if deadlineSet {
-		client.nc.SetWriteDeadline(time.Time{})
+	// Increment the flush pending signals if we are setting for the first time.
+	if _, ok := c.pcd[client]; !ok {
+		client.out.fsp++
 	}
-
 	client.mu.Unlock()
+
+	// Remember for when we return to the top of the loop.
 	c.pcd[client] = needFlush
```

Comment:
Is it safe outside the client lock? In sendOK() this is set under the client lock, and I see that you added a FIXME there. I think the idea of setting needsFlush there is because we call sendProto() with false, to indicate that this does not require a send in place, but ultimately, we want OK to be sent.

Reply:
Right. And any place we possibly touch this `pcd` map is from the client's readLoop go routine so we are ok.